### PR TITLE
Matmul grad

### PR DESCRIPTION
**Pull request by moskomule**
Hi, I add the matmul's gradient as discussed in the slack channel.

### CODE REVIEW

File: test_python/layers/test_matmul_train.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        def matmul(float(M,N) A, float(N,K) B) -> (output) {
+          output(i, j) +=! A(i, kk) * B(kk, j)
+        }
+        def matmul_grad(float(M,N) A, float(N,K) B, float(M,K) O_grad) -> (A_grad, B_grad){
+          A_grad(i, j) +=! O_grad(i, kk) * B(j, kk)
+          B_grad(i, j) +=! O_grad(kk, j) * A(kk, i)
+        }
+        """
+
+        # NOTE: TC doesn't support padding yet
```
Comment (prigoyal) [inline test_python/layers/test_matmul_train.py]: This note isnt relevant, can you remove this? :)

File: docs/source/framework/pytorch_integration/layers_database.rst [hunk 2]
```diff
@@ -184,6 +184,16 @@ Matmul
         output(i, j) +=! A(i, kk) * B(kk, j)
     }
 
+Matmul Gradient
+^^^^^^^^^^^^^^^
+
+.. code::
+
+    def matmul(float(M, N) A, float(N, K) B) -> (output) {
```
Comment (prigoyal) [inline docs/source/framework/pytorch_integration/layers_database.rst]: the name of the TC should be "matmul_grad" and the input O_grad is missing.

File: test_python/layers/test_matmul_train.py (hunk 1 above)
Comment (moskomule) [reply test_python/layers/test_matmul_train.py]: Yes, I forgot removing.

File: tensor_comprehensions/library/layers.yaml [hunk 3]
```diff
@@ -41,6 +41,12 @@
       output(i, j) +=! A(i, kk) * B(kk, j)
     }
```
Comment (prigoyal) [inline tensor_comprehensions/library/layers.yaml]: the new line is not needed and will break yaml parsing. see line 146-157 as an example :)

File: test_python/layers/test_matmul_train.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+        matmul = tc.define(MATMUL_LANG, name="matmul", training=True, backward="matmul_grad")
+        mat1 = Parameter(torch.randn(3, 4).cuda())
+        mat2 = Variable(torch.randn(4, 5).cuda(), requires_grad=True)
+        out = matmul(mat1, mat2, options=[tc.Options("mlp"), tc.Options("mlp")])
+        out.sum().backward()
+
+
+if __name__ == '__main__':
+unittest.main()
```
Comment (prigoyal) [inline test_python/layers/test_matmul_train.py]: indentation is missing here , this will fail to run

**Comment by prigoyal** [comment]
thanks for your contribution @moskomule :)
